ts/WoltLabSuite/Core/Api.ts

```diff
@@ -0,0 +1,294 @@
+    // the request handling has completed.
+    const showLoadingIndicator = this._showLoadingIndicator;
+    if (showLoadingIndicator) {
+      AjaxStatus.show();
+    }
+
+    try {
+      const response = await fetch(url, init);
+
+      if (!response.ok) {
+        const result = this.handleError(ErrorCode.STATUS_NOT_OK, response);
+
+        return Promise.reject(result);
+      }
+
+      const contentType = response.headers.get("content-type");
+      if (!contentType || !contentType.includes("application/json")) {
+        const result = this.handleError(ErrorCode.EXPECTED_JSON, response);
+
+        return Promise.reject(result);
+      }
+
+      let json: ResponseData;
+      try {
+        json = await response.json();
+      } catch (e) {
+        const result = this.handleError(ErrorCode.INVALID_JSON, response);
+
+        return Promise.reject(result);
+      }
+
+      // This is an explicit wait for the event loop to execute the
+      // callee function before we execute additional tasks.
+      return Promise.resolve(json.returnValues).then(async (result) => {
+        if (json.forceBackgroundQueuePerform) {
+          const backgroundQueue = await import("./BackgroundQueue");
+          backgroundQueue.invoke();
+        }
+        return result;
+      });
+    } catch (error) {
+      const result: ErrorResult = {
+        code: ErrorCode.CONNECTION_ERROR,
+        error,
+      };
+      let suppressError = false;
+      if (typeof this._failure === "function") {
+        suppressError = this._failure(result);
+      }
+
+      if (!suppressError) {
+        await this.genericError(result);
+      }
+
+      return Promise.reject(result);
+    } finally {
+      if (showLoadingIndicator) {
+        AjaxStatus.hide();
+      }
+
+      DomChangeListener.trigger();
+
+      // fix anchor tags generated through WCF::getAnchor()
+      document.querySelectorAll('a[href*="#"]').forEach((link: HTMLAnchorElement) => {
+        let href = link.href;
+        if (href.indexOf("AJAXProxy") !== -1 || href.indexOf("ajax-proxy") !== -1) {
+          href = href.substr(href.indexOf("#"));
+          link.href = document.location.toString().replace(/#.*/, "") + href;
+        }
+      });
+    }
+  }
+
+  private async handleError(code: ErrorCode, response: Response): Promise<ErrorResult> {
+    const result: ErrorResult = { code, response };
+    if (!this.suppressError(result)) {
+      await this.genericError(result);
+    }
+
+    return result;
+  }
+
+  private suppressError(result: ErrorResult): boolean {
+    if (typeof this._failure === "function") {
+      return this._failure(result);
+    }
+
+    return true;
+  }
+
+  private async genericError(result: ErrorResult): Promise<void> {
+    const html = await this.getErrorHtml(result);
+
+    if (html !== "") {
+      const uiDialog = await import("./Ui/Dialog");
+      const domUtil = await import("./Dom/Util");
+      const language = await import("./Language");
```

Comment:
One can assume that all of these are immediately available, but maybe use `Promise.all()` to load them in parallel?

Reply:
That makes sense, that code was refactored a few times but I missed this.